File: src/test/java/de/pflugradts/passbird/domain/model/transfer/CharValueTest.java
package de.pflugradts.passbird.domain.model.transfer;

import org.junit.jupiter.api.Test;

import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;

class CharValueTest {

    private static final String DIGITS = "0123456789";
    private static final String UPPER = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    private static final String LOWER = "abcdefghijklmnopqrstuvwxyz";
    private static final String SYMBOLS = "$&[{}(=*)+]!#/@\\';,.-~%`?^|\":<>_ ";

    @Test
    void shouldDetectDigits() {
        // given / when / then
        stream(DIGITS).forEach(c ->
            assertThat(CharValue.of(c).isDigit()).isTrue());

        stream(UPPER).forEach(c ->
            assertThat(CharValue.of(c).isDigit()).isFalse());
        stream(LOWER).forEach(c ->
            assertThat(CharValue.of(c).isDigit()).isFalse());
        stream(SYMBOLS).forEach(c ->
            assertThat(CharValue.of(c).isDigit()).isFalse());
    }

    @Test
    void shouldDetectUppercase() {
        // given / when / then
        stream(UPPER).forEach(c ->
            assertThat(CharValue.of(c).isUppercaseCharacter()).isTrue());

        stream(DIGITS).forEach(c ->
            assertThat(CharValue.of(c).isUppercaseCharacter()).isFalse());
        stream(LOWER).forEach(c ->
            assertThat(CharValue.of(c).isUppercaseCharacter()).isFalse());
        stream(SYMBOLS).forEach(c ->
            assertThat(CharValue.of(c).isUppercaseCharacter()).isFalse());
    }

    @Test
    void shouldDetectLowercase() {
        // given / when / then
        stream(LOWER).forEach(c ->
            assertThat(CharValue.of(c).isLowercaseCharacter()).isTrue());

        stream(DIGITS).forEach(c ->
            assertThat(CharValue.of(c).isLowercaseCharacter()).isFalse());
        stream(UPPER).forEach(c ->
            assertThat(CharValue.of(c).isLowercaseCharacter()).isFalse());
        stream(SYMBOLS).forEach(c ->
            assertThat(CharValue.of(c).isLowercaseCharacter()).isFalse());
    }

    @Test
    void shouldDetectSymbols() {
        // given / when / then
        stream(SYMBOLS).forEach(c ->
            assertThat(CharValue.of(c).isSymbol()).isTrue());

        stream(DIGITS).forEach(c ->
            assertThat(CharValue.of(c).isSymbol()).isFalse());
        stream(UPPER).forEach(c ->
            assertThat(CharValue.of(c).isSymbol()).isFalse());
        stream(LOWER).forEach(c ->
            assertThat(CharValue.of(c).isSymbol()).isFalse());
    }

    private Stream<Character> stream(String string) {
        return string.chars().mapToObj(i -> (char) i);
    }

}

File: src/test/kotlin/de/pflugradts/passbird/domain/model/transfer/CharValueTest.kt
package de.pflugradts.passbird.domain.model.transfer

import de.pflugradts.passbird.domain.model.transfer.CharValue.Companion.charValueOf
import org.junit.jupiter.api.Test
import strikt.api.expectThat
import strikt.assertions.isFalse
import strikt.assertions.isTrue

private const val DIGITS = "0123456789"
private const val UPPER = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
private const val LOWER = "abcdefghijklmnopqrstuvwxyz"
private const val SYMBOLS = "$&[{}(=*)+]!#/@\\';,.-~%`?^|\":<>_ "

class CharValueTest {
    @Test
    fun `should detect digits`() {
        // given / when / then
        DIGITS.toCharArray().forEach { expectThat(charValueOf(it).isDigit).isTrue() }
        DIGITS.toCharArray().forEach { expectThat(charValueOf(it).isUppercaseCharacter).isFalse() }
        DIGITS.toCharArray().forEach { expectThat(charValueOf(it).isLowercaseCharacter).isFalse() }
        DIGITS.toCharArray().forEach { expectThat(charValueOf(it).isAlphabeticCharacter).isFalse() }
        DIGITS.toCharArray().forEach { expectThat(charValueOf(it).isSymbol).isFalse() }
    }

    @Test
    fun `should detect uppercase`() {
        // given / when / then
        UPPER.toCharArray().forEach { expectThat(charValueOf(it).isDigit).isFalse() }
        UPPER.toCharArray().forEach { expectThat(charValueOf(it).isUppercaseCharacter).isTrue() }
        UPPER.toCharArray().forEach { expectThat(charValueOf(it).isLowercaseCharacter).isFalse() }
        UPPER.toCharArray().forEach { expectThat(charValueOf(it).isAlphabeticCharacter).isTrue() }
        UPPER.toCharArray().forEach { expectThat(charValueOf(it).isSymbol).isFalse() }
    }

    @Test
    fun `should detect lowercase`() {
        // given / when / then
        LOWER.toCharArray().forEach { expectThat(charValueOf(it).isDigit).isFalse() }
        LOWER.toCharArray().forEach { expectThat(charValueOf(it).isUppercaseCharacter).isFalse() }
        LOWER.toCharArray().forEach { expectThat(charValueOf(it).isLowercaseCharacter).isTrue() }
        LOWER.toCharArray().forEach { expectThat(charValueOf(it).isAlphabeticCharacter).isTrue() }
        LOWER.toCharArray().forEach { expectThat(charValueOf(it).isSymbol).isFalse() }
    }

    @Test
    fun `should detect symbols`() {
        // given / when / then
        SYMBOLS.toCharArray().forEach { expectThat(charValueOf(it).isDigit).isFalse() }
        SYMBOLS.toCharArray().forEach { expectThat(charValueOf(it).isUppercaseCharacter).isFalse() }
        SYMBOLS.toCharArray().forEach { expectThat(charValueOf(it).isLowercaseCharacter).isFalse() }
        SYMBOLS.toCharArray().forEach { expectThat(charValueOf(it).isAlphabeticCharacter).isFalse() }
        SYMBOLS.toCharArray().forEach { expectThat(charValueOf(it).isSymbol).isTrue() }
    }
}
